### PR TITLE
Fix output directory for copied byproducts

### PR DESCRIPTION
**Pull request by obsgolem**
CMake doesn't automatically append the config type to the output dir when using a custom command.

### CODE REVIEW

**Comment by jschwe** [comment]
I'm curious what the background of this PR is. Did you experience an issue with a multi-config generator?

**Comment by obsgolem** [comment]
Yes, I am having an issue where an executable is being copied to my `build` directory, not my `build/Debug` (or whatever) directory. This is with the VS generator. Some tests are failing, so I need to see if that is due to it assuming the behavior I was seeing, or if the behavior I was seeing is due to something complex I haven't found yet.

**Comment by obsgolem** [comment]
Apologies for the lockfile changes, don't know what is up with them. Want me to revert them?

**Review by jschwe — COMMENTED** [review]
> I am having an issue where an executable is being copied to my build directory, not my build/Debug (or whatever) directory. This is with the VS generator. Some tests are failing, so I need to see if that is due to it assuming the behavior I was seeing, or if the behavior I was seeing is due to something complex I haven't found yet.

What kind of tests are failing? Corrosions tests or tests from your project?
I guess changing the default location the artifact is copied to in Multi-Config mode has the potential to break user code, so we would need to document it in the release notes and also give a short reason what this change fixes.

File: test/output directory/output directory/CMakeLists.txt [hunk 1]
```diff
@@ -22,7 +22,7 @@ add_custom_command(TARGET cargo-build_rust_bin1 POST_BUILD
     COMMAND
     ${CMAKE_COMMAND} -E make_directory "${CMAKE_CURRENT_BINARY_DIR}/another_dir"
     COMMAND
-    ${CMAKE_COMMAND} -E copy_if_different "$<TARGET_PROPERTY:rust_bin1,LOCATION>" "${CMAKE_CURRENT_BINARY_DIR}/another_dir/moved_bin"
+    ${CMAKE_COMMAND} -E copy_if_different "$<TARGET_PROPERTY:rust_bin1,LOCATION_$<CONFIG>>" "${CMAKE_CURRENT_BINARY_DIR}/another_dir/moved_bin"
```
[the remark this answers is not in the document]
Comment (obsgolem) [reply test/output directory/output directory/CMakeLists.txt:25]: The CMake docs just have the general statement "generator expressions can be nested". As far as I am aware this pretty much always works. I can look for some examples in the stuff that ships with CMake if that would assuage doubts.

**Comment by obsgolem** [comment]
> What kind of tests are failing? Corrosions tests or tests from your project?
> I guess changing the default location the artifact is copied to in Multi-Config mode has the potential to break user code, so we would need to document it in the release notes and also give a short reason what this change fixes.

The output directory tests *were* failing, but after a couple rounds of revision I managed to fix that.

**Comment by obsgolem** [comment]
Failing test is not due to me I think

File: cmake/Corrosion.cmake [hunk 2]
```diff
@@ -100,7 +100,7 @@ function(_corrosion_set_imported_location_deferred target_name base_property out
         if(output_dir_curr_config)
             set(curr_out_dir "${output_dir_curr_config}")
         elseif(output_directory)
-            set(curr_out_dir "${output_directory}")
+            set(curr_out_dir "${output_directory}/${config_type}")
```
Comment (jschwe) [inline cmake/Corrosion.cmake:103]: I think we would need to check if `output_directory` contains a generator expression, and only when it **does not**, then append `${config_type}` to match the expected behavior (see e.g. https://cmake.org/cmake/help/latest/prop_tgt/RUNTIME_OUTPUT_DIRECTORY.html)

**Review by jschwe — CHANGES_REQUESTED** [review]
Looking at this again, I noticed that we unconditionally append the config_type now, when the expected behavior is to only append the config_type if the OUTPUT_DIRECTORY variable is not a generator expression

**Review by jschwe — APPROVED** [review]

**Comment by jschwe** [comment]
Thanks! And sorry for the long review times

**Comment by obsgolem** [comment]
Hey, if you can put up with long times between me responding to review, I can put up with long review times, thanks!

**Pull request merged**